REGRESSION (273213@main): ImageBitmap fails to report memory cost

kkinnunen-apple opened the pull request:
#### 6d1a61fa69dfe8e0a4b64ef6d629681068f37dd1
<pre>
REGRESSION (273213@main): ImageBitmap fails to report memory cost
<a href="https://bugs.webkit.org/show_bug.cgi?id=295618">https://bugs.webkit.org/show_bug.cgi?id=295618</a>
<a href="https://rdar.apple.com/155563791">rdar://155563791</a>

Reviewed by Matt Woodrow.

GC would not be triggered when many or large ImageBitmaps were
allocated.

Update ImageBitmap::m_memoryCost when creating the instance. This will
be used by bindings ReportExtraMemoryCost feature to report
the extra memory cost to the JS heap when the binding object is created
and visited.

Test: fast/images/imagebitmap-memory.html
* LayoutTests/fast/images/imagebitmap-memory-expected.html: Added.
* LayoutTests/fast/images/imagebitmap-memory.html: Added.
* Source/WebCore/html/ImageBitmap.cpp:
(WebCore::ImageBitmap::detach):
(WebCore::ImageBitmap::ImageBitmap):
(WebCore::ImageBitmap::~ImageBitmap):
(WebCore::ImageBitmap::takeImageBuffer):
(WebCore::ImageBitmap::memoryCost const):
(WebCore::ImageBitmap::updateMemoryCost): Deleted.
* Source/WebCore/html/ImageBitmap.h:

Canonical link: <a href="https://commits.webkit.org/300870@main">https://commits.webkit.org/300870@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/445765d8ea942682ef5f299d780b4a55c6f30f26

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124176 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43869 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34586 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131003 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/76268 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/db64b0e6-a9b5-4f6f-a4a6-ea568b149305) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/126053 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44610 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52469 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/94458 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/76268 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/09a6d3c7-842f-448b-b44e-0af12fd18782) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127130 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35548 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111074 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75049 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/627c7a36-553c-48ba-9297-6e005a267302) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/34492 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29236 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74489 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105289 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/29457 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133683 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51099 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38945 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/102932 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51485 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107292 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/102738 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26128 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48093 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26342 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/47988 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50960 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56735 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50407 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53759 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52083 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

mattwoodrow commented on Source/WebCore/html/ImageBitmap.cpp:
```diff
@@ -966,6 +966,7 @@ void ImageBitmap::createCompletionHandler(ScriptExecutionContext& scriptExecutio
 
 ImageBitmap::ImageBitmap(Ref<ImageBuffer> bitmap, bool originClean, bool premultiplyAlpha, bool forciblyPremultiplyAlpha)
     : m_bitmap(WTFMove(bitmap))
+    , m_memoryCost(m_bitmap->memoryCost())
```
It's not obvious to me why we need the atomic cached value, instead of just forwarding the query to the `ImageBuffer` on demand.

The original commit mentions the `ImageBuffer` not being threadsafe (which it is not), and we do have threadsafe refcounting despite that.

I don't think we actually have cross-thread usage of `ImageBuffer` though, we explicitly use the serialized-image-buffer concept to transfer ownership of data into a new thread-local `ImageBuffer` instance. It does feel like we should drop the thread safe refcounting, or move to a model where they are truly cross-thread objects.

Anyway, if you think this complexity is necessary here (or even not-provably-not-necessary), a comment would be useful imo :)

kkinnunen-apple replied:
It's general domain knowledge of GC related functions that GC might run on scavenger thread..

kkinnunen-apple replied:
Added a note to definition of the var: 
    std::atomic<size_t> m_memoryCost { 0 }; // Atomic, accessed from arbitrary thread by GC.

kkinnunen-apple committed:
REGRESSION (273213@main): ImageBitmap fails to report memory cost
https://bugs.webkit.org/show_bug.cgi?id=295618
rdar://155563791

Reviewed by Matt Woodrow.

GC would not be triggered when many or large ImageBitmaps were
allocated.

Update ImageBitmap::m_memoryCost when creating the instance. This will
be used by bindings ReportExtraMemoryCost feature to report
the extra memory cost to the JS heap when the binding object is created
and visited.

Test: fast/images/imagebitmap-memory.html
* LayoutTests/fast/images/imagebitmap-memory-expected.html: Added.
* LayoutTests/fast/images/imagebitmap-memory.html: Added.
* Source/WebCore/html/ImageBitmap.cpp:
(WebCore::ImageBitmap::detach):
(WebCore::ImageBitmap::ImageBitmap):
(WebCore::ImageBitmap::~ImageBitmap):
(WebCore::ImageBitmap::takeImageBuffer):
(WebCore::ImageBitmap::memoryCost const):
(WebCore::ImageBitmap::updateMemoryCost): Deleted.
* Source/WebCore/html/ImageBitmap.h:

Canonical link: https://commits.webkit.org/300870@main
